### PR TITLE
Fix issues with softmax layer in mixed precision distance metric computation

### PR DESCRIPTION
**Pull request by ofirgo**
* Fixed softmax node type in MP-related nodes mapping functions.
*  Fix KL divergence computation per softmax activation axis.
* Fix other missing types in node mapping by type functions for mixed precision.

See:
https://app.clickup.com/t/860r59zcb
https://app.clickup.com/t/32wz4zf

### CODE REVIEW

File: model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py [hunk 1]
```diff
@@ -283,7 +284,8 @@ def _compute_distance_matrix(self,
                                                   framework_attrs=self.interest_points[i].framework_attr,
                                                   compute_distance_fn=self.quant_config.compute_distance_fn)
 
-            distance_matrix[i] = point_distance_fn(baseline_tensors[i], mp_tensors[i], batch=True)
+            axis = self.interest_points[i].framework_attr.get(AXIS, None)
```
Comment (elad-c) [inline model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py:287]: get default is None. No nned to specify it

File: model_compression_toolkit/core/common/similarity_analyzer.py [hunk 2]
```diff
@@ -51,19 +51,24 @@ def _similarity_tensor_norm(x: np.ndarray, p: float = 2.0) -> np.ndarray:
...
         batch: Whether the similarity computation is per image or per tensor.
+        axis: Axis along which the operator has been computed.
 
     Returns: A flattened tensor which has the number of samples as is first dimension.
 
     """
 
-    if batch:
+    if axis is not None and batch:
+        f_t = t.reshape([t.shape[0], -1, t.shape[axis]])
```
Comment (elad-c) [inline model_compression_toolkit/core/common/similarity_analyzer.py:68]: this only works when "axis" is the last axis. need to permute the "axis" to the last index before reshaping like that

File: model_compression_toolkit/core/common/similarity_analyzer.py [hunk 3]
```diff
@@ -51,19 +51,24 @@ def _similarity_tensor_norm(x: np.ndarray, p: float = 2.0) -> np.ndarray:
...
+    elif axis is not None:
+        f_t = t.reshape([-1, t.shape[axis]])
```
Comment (elad-c) [inline model_compression_toolkit/core/common/similarity_analyzer.py:70]: same as above

File: model_compression_toolkit/core/common/similarity_analyzer.py [hunk 4]
```diff
@@ -186,6 +198,7 @@ def compute_lp_norm(float_tensor: np.ndarray,
         norm: whether to normalize the error function result.
         norm_eps: epsilon value for error normalization stability.
         batch: Whether to run batch similarity analysis or not.
+        axis: Axis along which the operator has been computed.
```
Comment (elad-c) [inline model_compression_toolkit/core/common/similarity_analyzer.py:201]: if you only added axis for the signiture to be the same, better write that it's not used.

File: model_compression_toolkit/core/keras/keras_implementation.py [hunk 5]
```diff
@@ -448,9 +449,10 @@ def get_node_distance_fn(self, layer_class: type,
                 return compute_kl_divergence
             elif node_type_name == SIGMOID:
                 return compute_cs
-        elif layer_class == tf.nn.softmax:
+        elif layer_class == tf.nn.softmax or layer_class == keras.layers.Softmax \
+                or (layer_class == TFOpLambda and SOFTMAX in framework_attrs[keras_constants.FUNCTION]):
             return compute_kl_divergence
-        elif layer_class == tf.nn.sigmoid:
+        elif layer_class == tf.nn.sigmoid or layer_class:
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/keras_implementation.py:455]: Are you sure this condition does what you meant it to do?

File: model_compression_toolkit/core/keras/keras_implementation.py [hunk 6]
```diff
@@ -417,8 +418,8 @@ def count_node_for_mixed_precision_interest_points(self, node: BaseNode) -> bool
             node_type_name = node.framework_attr[keras_constants.ACTIVATION]
             if node_type_name in [keras_constants.SOFTMAX, keras_constants.SIGMOID]:
                 return True
-        elif node.type in [tf.nn.softmax, tf.nn.sigmoid, Conv2D, DepthwiseConv2D, Conv2DTranspose, Dense, Concatenate,
-                           Add, tf.add]:
+        elif node.type in [tf.nn.softmax, keras.layers.Softmax, tf.nn.sigmoid, Conv2D, DepthwiseConv2D, Conv2DTranspose, Dense, Concatenate,
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/keras_implementation.py:421]: We usually use **tf.keras.layers.Softmax**, but it's the same class as **keras.layers.Softmax**.

File: tests/keras_tests/function_tests/test_sensitivity_metric_interest_points.py [hunk 7]
```diff
@@ -64,6 +67,17 @@ def dummy_representative_dataset():
     return ips, graph, fw_info
 
 
+def softmax_model(input_shape):
+    inputs = tf.keras.layers.Input(shape=input_shape)
+    x = tf.keras.layers.Conv2D(32, 4)(inputs)
+    x = tf.keras.layers.BatchNormalization()(x)
+    x = tf.keras.layers.Softmax(axis=-1)(x)
```
Comment (elad-c) [inline tests/keras_tests/function_tests/test_sensitivity_metric_interest_points.py:74]: try another axis value, not only -1

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py (hunk 1 above)
Comment (ofirgo) [reply model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py:287]: I know I just like it when it's more explicit this way, cause you know what to expect when the key doesn't exist, but I'll change it

File: model_compression_toolkit/core/keras/keras_implementation.py (hunk 5 above)
Comment (ofirgo) [reply model_compression_toolkit/core/keras/keras_implementation.py:455]: No, this was added by accident

**Review by elad-c — APPROVED** [review]
